add wordlist-based cleaning

zkamvar opened the pull request:
This will close #36 

This adds three new functions:

 - `clean_spelling()` will clean a specific vector based on a two-column wordlist. 
     + If the incoming vector is a factor, then the order of the wordlist is respected. 
     + a `.default` key can be defined to change all labels that aren't in the wordlist to a single default value (e.g. "unknown").
 - `clean_variable_spelling()` will clean all character/factor columns of a data frame based on a three-column wordlist, with the third column defining the columns to be checked.
     + If the third column contains the `.global` keyword, then the rows with this keyword will be implemented on all character/factor columns.
     + An additional numeric `sort_by` column can be defined to control the sorting of the wordlists. For example, missing data can have the value of `Inf` to ensure that it's always the last factor level.
 - `clean_variables()` will employ `clean_variable_labels()` and `clean_variable_spelling()` in that order.

And these new arguments:

 - `clean_data()` now also has `wordlist` (data frame or list of data frames), `spelling_vars` and `sort_by` to be passed onto `clean_variable_spelling()`.

AND A WHOLE LOTTA TESTS (92 to be exact)

I've also fixed a bug I introduced to `clean_dates()` where data were being returned as character if it could not be converted to a date... even if it was originally a factor. 

An example of the new `clean_data()` can be found in https://github.com/reconhub/linelist/pull/38#issuecomment-465053207

zkamvar committed:
add clean_spelling

zkamvar committed:
add clean_variable_spelling

 - add tests
 - add new test for clean_spelling

zkamvar committed:
add clean_variable spelling

 - tests are currently incomplete
 - there's a weird error with R CMD check messing with factor levels
   I'm printing them to be sure.
 - minimal documentation

zkamvar committed:
update clean_variable_spelling and clean_spelling

 - change dicts/dictionary to wordlists/wordlist
 - allow non-characters in clean_spelling
 - add tests for data frames
 - formalise ordering of factors in the end

The addresses the discussion in #36

zkamvar committed:
add explicit forcats imports; document

zkamvar committed:
add clean_variables

This function wraps both `clean_variable_labels()` and
`clean_variable_spelling()`. It will not initiate
`clean_variable_spelling()` unless there is a dictionary defined.

This further addresses #36

zkamvar committed:
change group to default to third column.

It was too easy to use the wordlist as a group wordlist.

 - change group = 3
 - throw error if numeric or character group is not in the wordlist
 - add warning for group = NULL

zkamvar committed:
fix and document logic for group column

zkamvar committed:
🤦🏽‍♂️

zkamvar committed:
add more logic and notes

zkamvar committed:
update documenation links [skip ci]

zkamvar commented:
Okay, @thibautjombart, It's ready:

## Using a wordlist for mis-spellings

<details>

<summary> Data Setup </summary>

``` r
set.seed(2019-02-19)
library("linelist")
toy_data <- messy_data(20)

messy_locations <- c("hopsital", "h\u00f4pital", "hospital", 
                     "m\u00e9dical", "clinic", 
                     "feild", "field", NA, "howdy")

toy_data$location <- factor(sample(messy_locations, 20, replace = TRUE))

wordlist <- data.frame(
  from  = c("hopsital", "hopital",  "medical", "feild", NA, ".default"),
  to = c("hospital", "hospital", "clinic",  "field", "unknown", "unknown"),
  variable = rep("location", 6),
  stringsAsFactors = FALSE
)

global_words <- data.frame(
  from = "not_a_case",
  to = "not a case",
  variable = ".global",
  stringsAsFactors = FALSE
)
```

</details>

With a data frame defining a wordlist, you can translate specific columns.

The wordlist can have a `.default` keyword:

``` r
wordlist
#>       from       to variable
#> 1 hopsital hospital location
#> 2  hopital hospital location
#> 3  medical   clinic location
#> 4    feild    field location
#> 5     <NA>  unknown location
#> 6 .default  unknown location
```

```r
clean_data4 <- clean_data(toy_data, wordlists = wordlist, error_tolerance = 0.8)
#> Warning: Unknown levels in `f`: field

head(toy_data)
#>      'ID Date of Onset. DisCharge.. GENDER_  Épi.Case_définition
#> 1 zzgyqk     2018-01-02  12/01/2018   Female          Not.a.Case
#> 2 nilpzb     2018-01-04  14/01/2018   female          Not.a.Case
#> 3 rmwama     2018-01-02  12/01/2018   Female            PROBABLE
#> 4 bfzoxo     2018-01-02  12/01/2018   FEMALE            probable
#> 5 atwapx     2018-01-11  21/01/2018     MALE          not a case
#> 6 inhwxl     2018-01-06  16/01/2018     male           confirmed
#>   messy/dates       lat      lon location
#> 1  2018-10-18  9.793946 48.98427  hôpital
#> 2  2018_10_17 10.826853 47.70421     <NA>
#> 3      female 13.346410 47.26488   clinic
#> 4        <NA> 16.910228 48.26746 hospital
#> 5  01-12-2001 14.309696 49.18833    howdy
#> 6  2018 10 19 11.601817 47.40893    howdy

head(clean_data4)
#>       id date_of_onset  discharge gender epi_case_definition messy_dates
#> 1 zzgyqk    2018-01-02 2018-01-12 female          not_a_case  2018-10-18
#> 2 nilpzb    2018-01-04 2018-01-14 female          not_a_case  2018-10-17
#> 3 rmwama    2018-01-02 2018-01-12 female            probable        <NA>
#> 4 bfzoxo    2018-01-02 2018-01-12 female            probable        <NA>
#> 5 atwapx    2018-01-11 2018-01-21   male          not_a_case  2001-12-01
#> 6 inhwxl    2018-01-06 2018-01-16   male           confirmed  2018-10-19
#>         lat      lon location
#> 1  9.793946 48.98427 hospital
#> 2 10.826853 47.70421  unknown
#> 3 13.346410 47.26488   clinic
#> 4 16.910228 48.26746 hospital
#> 5 14.309696 49.18833  unknown
#> 6 11.601817 47.40893  unknown
```

Factor levels respect the order of the wordlist

```r
toy_data$location
#>  [1] hôpital  <NA>     clinic   hospital howdy    howdy    médical 
#>  [8] clinic   hospital hopsital clinic   hospital howdy    hospital
#> [15] clinic   howdy    médical  clinic   howdy    howdy   
#> Levels: clinic hôpital hopsital hospital howdy médical

clean_data4$location # factor levels respect the order of the wordlist
#>  [1] hospital unknown  clinic   hospital unknown  unknown  clinic  
#>  [8] clinic   hospital hospital clinic   hospital unknown  hospital
#> [15] clinic   unknown  clinic   clinic   unknown  unknown 
#> Levels: hospital clinic unknown
```

You can append global variables with the `.global` keyword in the third column so that it applies to all character columns. This, however, will produce a lot of warnings.

``` r
global_words
#>         from         to variable
#> 1 not_a_case not a case  .global
wordlist2 <- rbind(wordlist, global_words, stringsAsFactors = FALSE)
head(clean_data(toy_data, wordlists = wordlist2, error_tolerance = 0.8))
```

<details>
<summary>WARNINGS</summary>

```r
#> Warning: Unknown levels in `f`: field, not a case
#> Warning in clean_spelling(x[[i]], d): None of the variables in x[[i]] were
#> found in d. Did you use the correct wordlist?
#> Warning: Unknown levels in `f`: not a case
#> Warning in clean_spelling(x[[i]], d): None of the variables in x[[i]] were
#> found in d. Did you use the correct wordlist?
#> Warning: Unknown levels in `f`: not a case
#> Warning in clean_spelling(x[[i]], d): None of the variables in x[[i]] were
#> found in d. Did you use the correct wordlist?
#> Warning: Unknown levels in `f`: not a case
#> Warning in clean_spelling(x[[i]], d): None of the variables in x[[i]] were
#> found in d. Did you use the correct wordlist?
#> Warning: Unknown levels in `f`: not a case
```

</details>

```r
#>       id date_of_onset  discharge gender epi_case_definition messy_dates
#> 1 zzgyqk    2018-01-02 2018-01-12 female          not a case  2018-10-18
#> 2 nilpzb    2018-01-04 2018-01-14 female          not a case  2018-10-17
#> 3 rmwama    2018-01-02 2018-01-12 female            probable        <NA>
#> 4 bfzoxo    2018-01-02 2018-01-12 female            probable        <NA>
#> 5 atwapx    2018-01-11 2018-01-21   male          not a case  2001-12-01
#> 6 inhwxl    2018-01-06 2018-01-16   male           confirmed  2018-10-19
#>         lat      lon location
#> 1  9.793946 48.98427 hospital
#> 2 10.826853 47.70421  unknown
#> 3 13.346410 47.26488   clinic
#> 4 16.910228 48.26746 hospital
#> 5 14.309696 49.18833  unknown
#> 6 11.601817 47.40893  unknown
```

<sup>Created on 2019-02-20 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

zkamvar commented:
Also, please excuse my French.

thibautjombart commented:
Looks great! Will review changes and get back to you interfrastically

thibautjombart commented on R/clean_spelling.R:
```diff
@@ -0,0 +1,96 @@
+    x <- as.character(x)
+  }
+
+  if (length(wordlist) < 2 || !is.data.frame(wordlist)) {
```
Will this work if the wordlist is provided as a matrix?

zkamvar replied:
No, but I can fix that :)

thibautjombart commented on R/clean_spelling.R:
```diff
@@ -0,0 +1,96 @@
+  dict <- dict[!is.na(dict)]
+  
+  # Recode data with forcats --------------------------------------------------
+  suppressWarnings(x <- forcats::fct_recode(x, !!!dict))
```
The `!!!` horrifies me. I think it's not not not okay. But I'm assuming `forcasts` does some useful stuff here (please tell me you didn't only choose to use it for the package's name?) ;)

zkamvar replied:
Fear is natural. This is directly from the documentation of `fct_recode()` since it takes a sequence of named elements (https://forcats.tidyverse.org/reference/fct_recode.html)

And no, I did not choose forcats simply because I love cats. It's actually a useful package because factors are nightmares.

thibautjombart commented on R/clean_variable_spelling.R:
```diff
@@ -0,0 +1,152 @@
+  if (is.data.frame(wordlists)) {
+
+    # There is a grouping column ----------------------------------------
+    if (!is.null(group) && length(group) == 1) {
```
I may well be missing it here, so do tell me off, but to me it looks like providing `group` isn't a much needed flexibility. Given that people will probably design their dicts after our examples, we can impose this. We could impose that `wordlist` is either:

- a named list of 2-col `data.frame`s in which case names are matched against column names in the cleaned dataset

- a single `data.frame` with 2 or 3 columns, and if there is a 3rd column it is assumed to be the variable names

Note that for matching variable names against the input datasete, match should be optional as one may use a general-purpose dictionary; but I suspect that's what the current code does

thibautjombart replied:
At any rate, I find `group` a bit misldeading a term. `variable_name`?

zkamvar replied:
I agree that `group` is misleading. I think I used it as a placeholder until a better name came up, but `variable_name` or `variables` or `spelling_variables` makes more sense.

I was designing this based on the dictionaries that I've seen for MSF in https://github.com/R4EPI/sitrep. Those dictionaries do not necessarily have the variable column as the third column.

zkamvar replied:
> Note that for matching variable names against the input datasete, match should be optional as one may use a general-purpose dictionary; but I suspect that's what the current code does

I'm not exactly sure what you mean here. Currently, there are two ways to go about this:

1. define the variables you want to work with
2. define a global dictionary

Regarding having a global dictionary, I think what you mentioned in https://github.com/reconhub/linelist/issues/36#issuecomment-465117954 is a good idea and will implement it.

Ultimately, what I want is for people to be able to use these wordlists for both spelling and defining label dictionaries with as little data manipulation on their side as possible.

zkamvar replied:
Implemented

thibautjombart replied:
Wikid

thibautjombart commented on R/guess_dates.R:
```diff
@@ -79,10 +81,13 @@ guess_dates <- function(x, error_tolerance = 0.1, first_date = NULL,
 
-  
+  # save the original x for later if nothing is converted
+  ox <- x
```
I thought that variable was for real, but then realised it's just an `ox`. Fine.

zkamvar replied:
wait... is this a pun on hoax?

thibautjombart commented on tests/testthat/test-clean_data.R:
```diff
@@ -5,14 +5,31 @@ DAT <- seq(as.Date("2018-01-01"), length.out = 10, by = "1 day")
+wordlist <- data.frame(
+  from  = c("hopsital", "hopital",  "medical", "feild"),
+  to    = c("hospital", "hospital", "clinic",  "field"),
+  var_shortname = rep("location", 4),
```
maybe just rename to `variable`? It's just `shortname` may suggest there's a `longname` too and get people confused.

zkamvar replied:
will do. I'll also show an example where there is a longname since that's what the MSF dictionaries use.

thibautjombart reviewed (changes requested):
This is great and is going to prove tremendously useful - thanks a lot! 

Just being picky here: do we need the `group` thing? I may be missing the point though. 

Will put another small feature request in the issue, as it should probably not be part of this review.

zkamvar committed:
change group to spelling_vars

This addresses https://github.com/reconhub/linelist/pull/38#discussion_r258018755

zkamvar committed:
allow matrices in clean_spelling

zkamvar committed:
add logic for default

zkamvar committed:
document and test .default key

zkamvar committed:
update cvs documentation

zkamvar committed:
add infastructure for .global spelling_var

zkamvar committed:
add documentation for .global variable

zkamvar committed:
add global wordlist to tests

zkamvar committed:
add another test for all global variables

thibautjombart commented:
Wonderful!
